kvserver: construct a write-only batch when DistinctSpans is true

sumeerbhola opened the pull request:
This fails a number of kvserver tests. I suspect there are cases
where DistinctSpans is true but some commands need to see the state
in an indexed batch (since not all commands were previously 
bypassing the batch by calling Distint()).

There is also the comment in api.pb.go that says:
Note that the calculation of distinct spans does not include 
intents in an EndTxnRequest.
which I don't quite follow.

I'd like opinions on whether it is worth pushing on this: an
indexed batch is 2-3x more expensive for set, because of the 
skip lists (pebble microbenchmarks), but I don't know
if it will show up as significant in real-world workloads.

Release note: None

sumeerbhola committed:
kvserver: construct a write-only batch when DistinctSpans is true

This fails a number of kvserver tests. I suspect there are cases
where DistinctSpans is true but some commands need to see the state
in an indexed batch (since not all commands were previously
bypassing the batch by calling Distint()).

There is also the comment in api.pb.go that says:
Note that the calculation of distinct spans does not include
intents in an EndTxnRequest.
which I don't quite follow.

I'd like opinions on whether it is worth pushing on this: an
indexed batch is 2-3x more expensive for set, because of the
skip lists (pebble microbenchmarks), but I don't know
if it will show up as significant in real-world workloads.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57661)
<!-- Reviewable:end -->

tbg commented:
The semantics of these optimized batches are, imo, terrible:

https://github.com/cockroachdb/cockroach/blob/8a449fc963b9c28b2844fc2eb1e7cd004495ff69/pkg/storage/engine.go#L662-L664

I would be very hesitant to build on top of them without a prior clean-up pass that results in more intuitive semantics and a reduced potential for bugs. Now I know you're basically taking the existing spaghetti and unifying it a bit, but can we do better? That this all works reliably (if it does) is not really enforced well and bugs here are a nightmare.

Right now, we have two such optimizations:

1. you can create a "write only batch", which, unfortunately, isn't actually write-only; it allows reads just fine except they won't see the batched writes.
2. we can derive a "Distinct()" ReadWriter from a batch which

        // only sees writes that
        // were performed before the Distinct batch was created. That is, the
        // returned batch will not read its own writes, but it will read writes to
        // the parent batch performed before the call to Distinct(), except if the
        // parent batch is a WriteOnlyBatch, in which case the Distinct() batch will
        // read from the underlying engine."

       This is all :spaghetti: and doesn't strike me as a sane foundation.

It's not entirely obvious how to make this all less error-prone, but one approach that could hold merit is one in which we're more explicit about the writes we're buffering. That is, instead of a ReadWriter, we pass the reader and writer around separately (in the lower levels of evaluation). Where we know that we don't need to read anything, we pass the engine (or a snapshot from it) as the reader, and a true WriteOnlyBatch (i.e. one that can't read) as a writer. If we decide later that we do need to read what we wrote, we have to flush the WriteOnlyBatch into a "real" batch and use that instead. This isn't "just going to work", it will need a bunch of refactoring.

sumeerbhola reviewed (commented):
I fully agree that the existing code is super confusing wrt `WriteOnlyBatch`, `Distinct`, reading from the `Engine` instead etc. I have bashed my head against it many times before I think I've finally achieved an understanding of what it does.

My intention in this WIP PR (which I should have stated up front) was to try to understand what are the underlying semantics we need in KV, and not to clean up the code. I would do the cleanup if I can get the WIP code to be correct first. The underlying semantics that are murky to me: when it is sufficient to use a non-indexed batch?

Currently we always create an indexed batch in `newBatchedEngine`, and for some commands (put/condition_put/init_put) we opt to construct a distinct batch which can read from the `engine` if the batch is not indexed. But the batch is always indexed, so these will see the state of the batch! Is `DistinctSpans` a sufficient condition to not need an indexed batch? The test failures suggest not. Can you help clarify what KV needs?
 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57661#-:-MO8CyN7AcWNcB26DyiX:b-30z04k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @petermattis, and @tbg)



<!-- Sent from Reviewable.io -->

tbg committed:
[dnm] kvserver: enforce !DistinctSpans for commit triggers

Commit triggers tend to need to read what they write since they both
write and resolve intents in one batch.

Release note: None

tbg commented:
Oh sorry! I hadn't understood the original intention here.

I took a look and pushed a commit to your branch (hope that's ok). The issue seems to be that we were setting `DistinctSpans` on the split trigger (and more generally, it can't ever be a good idea to set this for any commit trigger). Since a split both writes and resolve the intent for the right side of the split in the same batch, the resolution did not observe the write and thus we never actually write the right-hand side descriptor.

I am, more generally, worried that there are more bugs lurking with this code even if the tests go green. Would it be possible to "tee" the engine so that the writes go both to an unoptimized and an optimized batch? Any difference between the two would then indicate a bug. At the very least, the problem exposed for the split is general: intents can be written and resolved in one batch, so as it stands this PR will always result in lost writes. I think this translates to saying that we'd never want to employ this optimization, as it stands, for end transaction requests.

@nvanbenschoten might be a good person to opine on the inclusion of intents in the `DistinctSpans` optimization.

sumeerbhola reviewed (commented):
Thanks for looking and experimenting with this!

> Would it be possible to "tee" the engine so that the writes go both to an unoptimized and an optimized batch?

I think yes -- I'd prefer to do this after we have a clearer model for how this should work (as mentioned below).
 
> intents can be written and resolved in one batch, so as it stands this PR will always result in lost writes. I think this translates to saying that we'd never want to employ this optimization, as it stands, for end transaction requests. @nvanbenschoten might be a good person to opine on the inclusion of intents in the DistinctSpans optimization.

It would be really good to clean up our model of how this should work. I am worried that the code thinks it is using the write-only batch optimization, and won't see the mutations when in-fact it is usually an indexed batch -- things are working, but fragile, and we are leaving optimizations on the table. I am assuming end-transaction is not the only issue since a number of tests are failing even after your fix to that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57661#-:-MOhOcTG-gMnCA9fE_tG:b-gqb466)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @petermattis, and @tbg)



<!-- Sent from Reviewable.io -->

tbg committed:
kvserver: add test demonstrating lost writes

Add an example test (TestFoo) that shows how we lose writes with
the distinct optimization. The txn

1. writes at \xff (just to anchor the txn) without pipelining
2. batch(put(\xff\x00), commit)

The batch in 2) will have DistinctSpans set and will move the txn record
directly to committed (without going through staging). It fails however
to resolve the intent at \xff\x00 which thus becomes abortable once the
txn record is gone, which also happens eagerly.

Release note: None

tbg committed:
kvserver: hacky fix for TestFoo

... and possibly the other test failures. Use `DistinctSpans` only when
the batch contains entirely of allowlisted requests, and only if the
optionally present EndTxn does not contain lock spans overlapping the
batch (as these may end up being resolved synchronously with the commit,
which would be buggy without read-what-you-wrote).

This isn't production-grade code by a long shot, though it does get
the job done, when performance is not a concern.

Release note: None

tbg commented:
> I think yes -- I'd prefer to do this after we have a clearer model for how this should work (as mentioned below).

I was coming from the same angle, in seeing more violations I was hoping for a cheap way to build a mental model. But I think I have a basic understanding now. 

> I am assuming end-transaction is not the only issue since a number of tests are failing even after your fix to that.

I think it's all about EndTransaction still. Remember that I only fixed the commit trigger case, but mentioned:

> At the very least, the problem exposed for the split is general: intents can be written and resolved in one batch, so as it stands this PR will always result in lost writes.

I added a commit with a unit test that demonstrates exactly this. Possible that after fixing that there are still test failures here, but then that would indicate a gap in my understanding. Unfortunately it's not a one-line fix so I haven't tried yet.

I think our use of DistinctSpans was just always wrong. The - I believe - original use was for this `optimizePuts` optimization:

https://github.com/cockroachdb/cockroach/blob/b38eaf6331fa8c397a523ed3c5ecf78baa272abd/pkg/kv/kvserver/replica_evaluate.go#L43-L59

which also uses `DistinctSpans`, though I think it uses it correctly: if DistinctSpans is true, it trusts that all of the puts are distinct and so only needs to check whether their underlying keyspace is empty (so that it can turn the puts "blind"). Otherwise, it does its own deduplication to optimize at least a prefix of the batch (though that might be premature optimization).

I can only guess (though archaeology could confirm) that we later also started optimizing the readWriter based on DistinctSpans, though... incorrectly. It seems to me that we want to rename `DistinctSpans` to something else that indicates that the batch will evaluate without ever reading what it wrote. Also, the optimization should only apply to batches that have:

1. allowlist-based approach: only cput, put, delete, deleterange, committing end transaction w/o special flags (triggers etc), and
2. if endtransaction, lock spans not overlapping the request spans.

tbg commented:
I just added a commit implementing 1+2 above. It does fix TestFoo. It may also fix the remaining test issues, let's see. If it does, there is a decent amount of confidence that this is the only problem here.

sumeerbhola commented:
There are a lot of ignored kvserver tests https://teamcity.cockroachdb.com/viewLog.html?buildId=2526866&buildTypeId=Cockroach_UnitTests
Is that expected?

tbg commented:
Not desired, but expected.

sumeerbhola commented:
> If it does, there is a decent amount of confidence that this is the only problem here.

The tests are passing. Presumably you still want me to try the "tee the engine so that the writes go both to an unoptimized and an optimized batch"?

@nvanbenschoten do you have any thoughts on this?

tbg commented:
> Presumably you still want me to try the "tee the engine so that the writes go both to an unoptimized and an optimized batch"

Mhm, I would feel better about it, but I think conceptually we are mostly in the clear here. Nathan is out, so am I starting tomorrow, so we may have to think this through next year. But - you've definitely found a great mess waiting to happen, I am very thankful for your poking around here.

nvanbenschoten commented:
> I think our use of DistinctSpans was just always wrong. The - I believe - original use was for this optimizePuts optimization:
>
> ...
>
>which also uses DistinctSpans, though I think it uses it correctly: if DistinctSpans is true, it trusts that all of the puts are distinct and so only needs to check whether their underlying keyspace is empty (so that it can turn the puts "blind"). Otherwise, it does its own deduplication to optimize at least a prefix of the batch (though that might be premature optimization).
>
>I can only guess (though archaeology could confirm) that we later also started optimizing the readWriter based on DistinctSpans, though... incorrectly.

This seems to be correct. We originally introduced this field in https://github.com/cockroachdb/cockroach/commit/3c815aed88f51e21f5bfa1fb2c02819776a83d9e, where it was only set on batches with an EndTxn, but indicated that all write requests _other than_ the EndTxn were to distinct keys. We then started using it in https://github.com/cockroachdb/cockroach/commit/49dc611189381e22cdfe884ecfa6aa1d8a084e8f, where we created a new `Distinct` batch for each of the writes _other than_ the EndTxn when the `DistinctSpans` flag was set.

At first, I also thought this was broken, but I actually think it works in a crazy way and only because a distinct batch means something different than what I thought it meant. A distinct batch doesn't require that writes to it are distinct. It also doesn't make a claim that the batch is write-only. Instead, it only makes a claim that the batch won't provide read-your-writes semantics. So while very confusing, it's fine that we use distinct batches for the writes as long as we don't for the EndTxn itself, which may need to read the writes during intent resolution.

My first suggestion here would be to abandon the term "distinct batch" and replace it with "non-indexed batch". This will eliminate the mismatch between CRDB and Pebble and make the meaning of this more clear (or at least less easy to confidently misunderstand). We can then draw a clearer distinction between "batches", "non-indexed batches", and "write-only batches" (which are non-indexed batches that don't allow reads). That's still slightly inverting what pebble provides ("batches" and "indexed batches"), but it still seems like a step in the right direction.

Once we do that, I think we can continue in the spirit of this PR. I think the original idea (correct me if I'm wrong) was to create a single non-indexed batch for the entirety of a "distinct" `roachpb.BatchRequest`. And so not a "write-only" batch as the PR title says, but a "non-indexed" batch. But that runs into issues with the EndTxn of those batches because they need to read the intents written by the other requests. One possibility here is that we only create a non-indexed batch for the entirely of a "distinct" `roachpb.BatchRequest` if it evaluates on the 1PC path. This avoids the question of ever needing to read-your-writes on the batch or ever needing to flush a non-indexed batch back into a parent indexed batch. What do you think?

sumeerbhola reviewed (commented):
Thanks for the comments @tbg and @nvanbenschoten 
I am re-reading this PR, and the comments here and the related PRs where this saga started, to form a better mental model, and have some more questions below.
Let me know if these questions below are easier to discuss synchronously and I'll setup some time to chat.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57661#-:-MQiccvI9E7fGSZFnDYU:b-yoppm7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @petermattis, and @tbg)

___
*[pkg/kv/kvserver/replica_send.go, line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57661#-MQZVBmY43dGp3BuazfN:-MQZVBmZ2CKDYPIx-MuP:bc3on6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/6046ae141abc63bc2759969ebe3dcce74e615490/pkg/kv/kvserver/replica_send.go#L97)):*
> ```Go
> 				spans = append(spans, t.Header().Span())
> 			case *roachpb.EndTxnRequest:
> 				spans = append(spans, t.LockSpans...)
> ```

Do these `LockSpans` represent the writes prior to this batch, or also include those in this batch? If it didn't include the ones in this batch, we could have `DistinctSpans=true` below, even though this batch needs to read the intents written, so I am guessing it does include the ones in this batch. So the only case an `EndTxnRequest` can end up in a batch with `DistinctSpans=true` is if there are no other requests in it. The code in `maybeStripInFlightWrites` also suggests this.
Is this all correct?

I am a little confused about the different places `DistinctSpans` is set, modified and subsequently used.
- Set or modified: Currently it is in set in `txnCommitter.SendLocked`, I think only when the BatchRequest has an EndTxn. This looks like client-side. Then there is the server-side code in `maybeStripInFlightWrites` which also only executes when the `BatchRequest` has an EndTxn. I don't see any other place. Are we not setting it for non-EndTxn BatchRequests, and if yes is this an oversight?
  And then there is the modification here, which looks server-side. I am tempted to hoist this where we initially set `DistinctSpans` on the server-side if there was such a place.

- Used: There is the use where `Distinct()` is called on the Batch. Setting that aside, there is the call to `optimizePuts` which can avoid the map overhead if `DistinctSpans` was set. This use case is using weaker semantics since it uses `DistinctSpans` in conjunction with looking only at a prefix of puts/cputs/initputs in the BatchRequest. And IIUC this is happening after the modification in `sendWithRangeID` which means the stronger semantics will force more code to use the map. Is my understanding correct?

This leads to another question. Based on reading the code in @nvanbenschoten last comment, it seems that the original semantics of `DistinctSpans` did not need to apply to the whole `BatchRequest`, since it was primarily an optimization for when to flush writes buffered in Go across the cgo boundary into the rocksdb batch, at which point they became visible to other requests that included the batch in their iterator. Some requests in the `BatchRequest` did not need to read what was in the batch so could use a `DistinctBatch`, while others could force the flush and see everything. We are now trying to repurpose this towards constructing a batch that will not expose its state to *any* of the requests in the `BatchRequest`, which requires a stricter guarantee. Do you have a sense whether this will be a useful optimization, or will `DistinctSpans` will usually end up being false? I can also try to measure it if you point me to representative workloads to run.


 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
Thanks for continuing to push on this! We seem to be getting closer to some clarity and some much cleaner semantics around these non-indexed batches and their use in KV.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57661#-:-MQiv8wl76TkHKN3gGrp:b55m1oz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @petermattis, @sumeerbhola, and @tbg)

___
*[pkg/kv/kvserver/replica_send.go, line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57661#-MQZVBmY43dGp3BuazfN:-MQirkDD6-pKHqtTvgBG:b-hkg4ih) ([raw file](https://github.com/cockroachdb/cockroach/blob/6046ae141abc63bc2759969ebe3dcce74e615490/pkg/kv/kvserver/replica_send.go#L97)):*
> Do these LockSpans represent the writes prior to this batch, or also include those in this batch?

They include both.

> So the only case an EndTxnRequest can end up in a batch with DistinctSpans=true is if there are no other requests in it.

I don't think this is true. I think `DistinctSpans` specifically doesn't look at the writes that the EndTxn is performing. That's what the "Note that the calculation of distinct spans does not include intents in an EndTxnRequest." comment is indicating. So this is why the flag can be set when a batch contains writes and an EndTxn. But it's also why the EndTxn itself doesn't use a `Distinct` batch.

> Are we not setting it for non-EndTxn BatchRequests, and if yes is this an oversight?

No, we aren't unfortunately. See this TODO, which seems to have been lost somewhere in the refactoring process: https://github.com/cockroachdb/cockroach/commit/3c815aed88f51e21f5bfa1fb2c02819776a83d9e#diff-230152fb9c39bac311935d3cda2ec1d43791a94de9316a1366b314b21337f51eR344

I don't think the intention was ever to limit this to batches that contain EndTxn requests, but the generalization was never made.

>  We are now trying to repurpose this towards constructing a batch that will not expose its state to any of the requests in the BatchRequest, which requires a stricter guarantee. Do you have a sense whether this will be a useful optimization, or will DistinctSpans will usually end up being false? 

Yeah, this matches my understanding as well. I can't say definitively whether this will be a useful optimization or whether it will undermine the entire thing. It will certainly prevent the optimization from applying in some cases, like batches that contain `Put, EndTxn` at the end of an ongoing transaction. However, it may apply to the 1-phase commit fast-path (where the EndTxn is stripped on the server and writes are never written as intents), in which case it may improve things for a large number of workloads. It would be worth testing things out with `kv0` (with different `--batch` sizes) if you're able to put a prototype together.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57661)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis and @tbg)

___
*[pkg/kv/kvserver/replica_send.go, line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57661#-MQZVBmY43dGp3BuazfN:-MQlrsc1C2X5DzLDsd2Z:b-n18cwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/6046ae141abc63bc2759969ebe3dcce74e615490/pkg/kv/kvserver/replica_send.go#L97)):*
I don't have a good understanding of the details of our "end-transaction" handling, hence these questions. Is there something I can read, code or docs (I have read https://github.com/cockroachdb/cockroach/blob/master/docs/tech-notes/txn_coord_sender.md), before asking more questions?

>  I think DistinctSpans specifically doesn't look at the writes that the EndTxn is performing. That's what the "Note that the calculation of distinct spans does not include intents in an EndTxnRequest." comment is indicating. So this is why the flag can be set when a batch contains writes and an EndTxn.

I was specifically referring to the change by @tbg here. It is adding the `LockSpans` to `spans` and since those will contain all puts that have happened in this txn, I think that would mean there is overlap when calling `MergeSpans`, unless there are no puts in this batch.

The other thing I am somewhat unclear about is why `EndTxn` is reading intents in the 2PC case. My mental model is that there are two cases:
- Parallel commit: intent resolution will not happen in this batch and the transaction enters STAGING state. Intents will not be read. But to return success to the client we need to prove the intents, `InFlightWrites`,  have been written. I am guessing this proving just requires waiting for success and not actually reading those intents.
- Transaction goes straight to COMMITTED. This is when there are no `InFlightWrites`. The EndTxn will also do intent resolution, and needs to read the intents.

Is this correct?   


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57661)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis and @tbg)

___
*[pkg/kv/kvserver/replica_send.go, line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57661#-MQZVBmY43dGp3BuazfN:-MQxB7R9-NEEcBQS1Q_D:b-mekr54) ([raw file](https://github.com/cockroachdb/cockroach/blob/6046ae141abc63bc2759969ebe3dcce74e615490/pkg/kv/kvserver/replica_send.go#L97)):*
> Is there something I can read, code or docs (I have read https://github.com/cockroachdb/cockroach/blob/master/docs/tech-notes/txn_coord_sender.md), before asking more questions?

The commentary on `txnCommitter` might be helpful here. And `txnPipeliner.attachLocksToEndTxn`. Also, comments on `resolveLocalLocks`.

> Parallel commit: intent resolution will not happen in this batch and the transaction enters STAGING state. Intents will not be read. But to return success to the client we need to prove the intents, InFlightWrites, have been written. I am guessing this proving just requires waiting for success and not actually reading those intents.

This is almost all correct. The only clarification I would make is that the act of providing the success of intents is done by `QueryIntent` operations, which are split but concurrent with the EndTxn. See the commentary on `txnPipeliner` for more.

> Transaction goes straight to COMMITTED. This is when there are no InFlightWrites. The EndTxn will also do intent resolution, and needs to read the intents.

Right. It will perform intent resolution for all intents on the same range as the transaction record.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
@tbg  @nvanbenschoten 
Before I start cleaning this up, I want to confirm two things:
- How this optimization would be applied
- Whether/when it would be beneficial

In https://github.com/sumeerbhola/cockroach/commit/eeeec51bd40ef47e743dc0c9ca47cf15710bae09 I've adapted @tbg's code to do the computation of whether an un-indexed batch can be used, in `Replica`. It doesn't look at the existing value of `DistinctSpans`set by the client, since (a) the batch may have been split, and (b) the current computation of `DistinctSpans` is only done for an `EndTxn`. I don't want to touch the use of `DistinctSpans` (though will rename it) since its weaker semantics are used in `optimizePuts`.
Do you think doing such a computation over spans is too expensive, or is it fine as a starting point (of course subject to running benchmarks)? If too expensive, are there any better options?

I ran this with
`kv95/enc=false/nodes=1` and the log statements show only a tiny fraction of BatchRequests with writes can use this optimization e.g. log statement
``` WriteBatch: Count: 200000, UnIndexed: 4452```

And I ran with `kv95limited-spans/enc=false/nodes=1/splt=0/seq` from https://github.com/cockroachdb/cockroach/pull/58670/commits/2a0ea0939d0ba0196a9aba787841fa354591a8aa which does the write in a transaction which contains a select for update followed by an upsert, and the fraction that benefits is very high
e.g. WriteBatch: Count: 48500, UnIndexed: 48446
I am guessing the first one will be an 1PC transaction and the second one is not. I need some help in understanding why the first one is so bad. I am happy to add more instrumentation and rerun.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57661#-:-MR6uTGr-tUq3C4C7ZhV:b4x0gl2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis and @tbg)



<!-- Sent from Reviewable.io -->
